// @sentry/nestjs/build/esm/decorators.js
import { captureException, SEMANTIC_ATTRIBUTE_SENTRY_OP, SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN } from '@sentry/core';
import * as Sentry from '@sentry/node';
import { startSpan } from '@sentry/node';
import { isExpectedError } from './helpers.js';

const SentryCron = (monitorSlug, monitorConfig) => {
  return (target, propertyKey, descriptor) => {
    const originalMethod = descriptor.value;
    descriptor.value = function(...args) {
      return Sentry.withMonitor(
        monitorSlug,
        () => {
          return originalMethod.apply(this, args);
        },
        monitorConfig
      );
    };
    copyFunctionNameAndMetadata({ originalMethod, descriptor });
    return descriptor;
  };
};
function SentryTraced(op = "function") {
  return function(_target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function(...args) {
      return startSpan(
        {
          op,
          name: propertyKey,
          attributes: {
            [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.function.nestjs.sentry_traced",
            [SEMANTIC_ATTRIBUTE_SENTRY_OP]: op
          }
        },
        () => {
          return originalMethod.apply(this, args);
        }
      );
    };
    copyFunctionNameAndMetadata({ originalMethod, descriptor });
    return descriptor;
  };
}
function SentryExceptionCaptured() {
  return function(target, propertyKey, descriptor) {
    const originalCatch = descriptor.value;
    descriptor.value = function(exception, host, ...args) {
      if (isExpectedError(exception)) {
        return originalCatch.apply(this, [exception, host, ...args]);
      }
      captureException(exception, { mechanism: { handled: false, type: "auto.function.nestjs.exception_captured" } });
      return originalCatch.apply(this, [exception, host, ...args]);
    };
    copyFunctionNameAndMetadata({ originalMethod: originalCatch, descriptor });
    return descriptor;
  };
}
function copyFunctionNameAndMetadata({
  originalMethod,
  descriptor
}) {
  Object.defineProperty(descriptor.value, "name", {
    value: originalMethod.name,
    configurable: true,
    enumerable: true,
    writable: true
  });
  if (typeof Reflect !== "undefined" && typeof Reflect.getMetadataKeys === "function") {
    const originalMetaData = Reflect.getMetadataKeys(originalMethod);
    for (const key of originalMetaData) {
      const value = Reflect.getMetadata(key, originalMethod);
      Reflect.defineMetadata(key, value, descriptor.value);
    }
  }
}

export { SentryCron, SentryExceptionCaptured, SentryTraced };
