// @sentry/nestjs/build/esm/setup.js
import { Injectable, Catch, Global, Module, Logger, HttpException } from '@nestjs/common';
import { APP_INTERCEPTOR, BaseExceptionFilter } from '@nestjs/core';
import { getIsolationScope, getDefaultIsolationScope, debug, captureException } from '@sentry/core';
import { isExpectedError } from './helpers.js';

class SentryTracingInterceptor {
  constructor() {
    this.__SENTRY_INTERNAL__ = true;
  }
  /**
   * Intercepts HTTP requests to set the transaction name for Sentry tracing.
   */
  intercept(context, next) {
    if (getIsolationScope() === getDefaultIsolationScope()) {
      debug.warn("Isolation scope is still the default isolation scope, skipping setting transactionName.");
      return next.handle();
    }
    if (context.getType() === "http") {
      const req = context.switchToHttp().getRequest();
      if ("routeOptions" in req && req.routeOptions?.url) {
        getIsolationScope().setTransactionName(`${(req.method || "GET").toUpperCase()} ${req.routeOptions.url}`);
      } else if ("route" in req && req.route?.path) {
        getIsolationScope().setTransactionName(`${(req.method || "GET").toUpperCase()} ${req.route.path}`);
      }
    }
    return next.handle();
  }
}
Injectable()(SentryTracingInterceptor);
class SentryGlobalFilter extends BaseExceptionFilter {
  constructor(applicationRef) {
    super(applicationRef);
    this.__SENTRY_INTERNAL__ = true;
    this._logger = new Logger("ExceptionsHandler");
  }
  /**
   * Catches exceptions and reports them to Sentry unless they are expected errors.
   */
  catch(exception, host) {
    const contextType = host.getType();
    if (contextType === "graphql") {
      if (exception instanceof HttpException) {
        throw exception;
      }
      if (exception instanceof Error) {
        this._logger.error(exception.message, exception.stack);
      }
      captureException(exception, {
        mechanism: {
          handled: false,
          type: "auto.graphql.nestjs.global_filter"
        }
      });
      throw exception;
    }
    if (contextType === "rpc") {
      if (exception instanceof HttpException) {
        throw exception;
      }
      if (!(exception instanceof Error)) {
        if (!isExpectedError(exception)) {
          captureException(exception, {
            mechanism: {
              handled: false,
              type: "auto.rpc.nestjs.global_filter"
            }
          });
        }
        throw exception;
      }
      if (!isExpectedError(exception)) {
        captureException(exception, {
          mechanism: {
            handled: false,
            type: "auto.rpc.nestjs.global_filter"
          }
        });
      }
      this._logger.warn(
        "IMPORTANT: RpcException should be handled with a dedicated Rpc exception filter, not the generic SentryGlobalFilter"
      );
      this._logger.error(exception.message, exception.stack);
      return;
    }
    if (!isExpectedError(exception)) {
      captureException(exception, {
        mechanism: {
          handled: false,
          type: "auto.http.nestjs.global_filter"
        }
      });
    }
    return super.catch(exception, host);
  }
}
Catch()(SentryGlobalFilter);
class SentryModule {
  /**
   * Configures the module as the root module in a Nest.js application.
   */
  static forRoot() {
    return {
      module: SentryModule,
      providers: [
        {
          provide: APP_INTERCEPTOR,
          useClass: SentryTracingInterceptor
        }
      ]
    };
  }
}
Global()(SentryModule);
Module({
  providers: [
    {
      provide: APP_INTERCEPTOR,
      useClass: SentryTracingInterceptor
    }
  ]
})(SentryModule);

export { SentryGlobalFilter, SentryModule };
