// @sentry/nestjs/build/cjs/integrations/helpers.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@sentry/core');

const sentryPatched = "sentryPatched";
function isPatched(target) {
  if (target.sentryPatched) {
    return true;
  }
  core.addNonEnumerableProperty(target, sentryPatched, true);
  return false;
}
function getMiddlewareSpanOptions(target, name = void 0, componentType = void 0) {
  const span_name = name ?? target.name;
  const origin = componentType ? `auto.middleware.nestjs.${componentType}` : "auto.middleware.nestjs";
  return {
    name: span_name,
    attributes: {
      [core.SEMANTIC_ATTRIBUTE_SENTRY_OP]: "middleware.nestjs",
      [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: origin
    }
  };
}
function getEventSpanOptions(event) {
  return {
    name: `event ${event}`,
    attributes: {
      [core.SEMANTIC_ATTRIBUTE_SENTRY_OP]: "event.nestjs",
      [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.event.nestjs"
    },
    forceTransaction: true
  };
}
function getBullMQProcessSpanOptions(queueName) {
  return {
    name: `${queueName} process`,
    attributes: {
      [core.SEMANTIC_ATTRIBUTE_SENTRY_OP]: "queue.process",
      [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.queue.nestjs.bullmq",
      "messaging.system": "bullmq",
      "messaging.destination.name": queueName
    },
    forceTransaction: true
  };
}
function instrumentObservable(observable, activeSpan) {
  if (activeSpan) {
    observable.subscribe = new Proxy(observable.subscribe, {
      apply: (originalSubscribe, thisArgSubscribe, argsSubscribe) => {
        return core.withActiveSpan(activeSpan, () => {
          const subscription = originalSubscribe.apply(thisArgSubscribe, argsSubscribe);
          subscription.add(() => activeSpan.end());
          return subscription;
        });
      }
    });
  }
}
function getNextProxy(next, span, prevSpan) {
  return new Proxy(next, {
    apply: (originalNext, thisArgNext, argsNext) => {
      span.end();
      if (prevSpan) {
        return core.withActiveSpan(prevSpan, () => {
          return Reflect.apply(originalNext, thisArgNext, argsNext);
        });
      } else {
        return Reflect.apply(originalNext, thisArgNext, argsNext);
      }
    }
  });
}

exports.getBullMQProcessSpanOptions = getBullMQProcessSpanOptions;
exports.getEventSpanOptions = getEventSpanOptions;
exports.getMiddlewareSpanOptions = getMiddlewareSpanOptions;
exports.getNextProxy = getNextProxy;
exports.instrumentObservable = instrumentObservable;
exports.isPatched = isPatched;
