// @sentry/nestjs/build/cjs/integrations/sentry-nest-event-instrumentation.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const instrumentation = require('@opentelemetry/instrumentation');
const core = require('@sentry/core');
const helpers = require('./helpers.js');

const supportedVersions = [">=2.0.0"];
const COMPONENT = "@nestjs/event-emitter";
class SentryNestEventInstrumentation extends instrumentation.InstrumentationBase {
  constructor(config = {}) {
    super("sentry-nestjs-event", core.SDK_VERSION, config);
  }
  /**
   * Initializes the instrumentation by defining the modules to be patched.
   */
  init() {
    const moduleDef = new instrumentation.InstrumentationNodeModuleDefinition(COMPONENT, supportedVersions);
    moduleDef.files.push(this._getOnEventFileInstrumentation(supportedVersions));
    return moduleDef;
  }
  /**
   * Wraps the @OnEvent decorator.
   */
  _getOnEventFileInstrumentation(versions) {
    return new instrumentation.InstrumentationNodeModuleFile(
      "@nestjs/event-emitter/dist/decorators/on-event.decorator.js",
      versions,
      (moduleExports) => {
        if (instrumentation.isWrapped(moduleExports.OnEvent)) {
          this._unwrap(moduleExports, "OnEvent");
        }
        this._wrap(moduleExports, "OnEvent", this._createWrapOnEvent());
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports, "OnEvent");
      }
    );
  }
  /**
   * Creates a wrapper function for the @OnEvent decorator.
   */
  _createWrapOnEvent() {
    return function wrapOnEvent(original) {
      return function wrappedOnEvent(event, options) {
        const decoratorResult = original(event, options);
        return (target, propertyKey, descriptor) => {
          if (!descriptor.value || typeof descriptor.value !== "function" || target.__SENTRY_INTERNAL__ || // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          descriptor.value.__SENTRY_INSTRUMENTED__) {
            return decoratorResult(target, propertyKey, descriptor);
          }
          function eventNameFromEvent(event2) {
            if (typeof event2 === "string") {
              return event2;
            } else if (Array.isArray(event2)) {
              return event2.map(eventNameFromEvent).join(",");
            } else return String(event2);
          }
          const originalHandler = descriptor.value;
          const handlerName = originalHandler.name || propertyKey;
          let eventName = eventNameFromEvent(event);
          descriptor.value = async function(...args) {
            if (Reflect.getMetadataKeys(descriptor.value).includes("EVENT_LISTENER_METADATA")) {
              const eventData = Reflect.getMetadata("EVENT_LISTENER_METADATA", descriptor.value);
              if (Array.isArray(eventData)) {
                eventName = eventData.map((data) => {
                  if (data && typeof data === "object" && "event" in data && data.event) {
                    return eventNameFromEvent(data.event);
                  }
                  return "";
                }).reverse().join("|");
              }
            }
            return core.withIsolationScope(() => {
              return core.startSpan(helpers.getEventSpanOptions(eventName), async () => {
                try {
                  const result = await originalHandler.apply(this, args);
                  return result;
                } catch (error) {
                  core.captureException(error, {
                    mechanism: {
                      handled: false,
                      type: "auto.event.nestjs"
                    }
                  });
                  throw error;
                }
              });
            });
          };
          descriptor.value.__SENTRY_INSTRUMENTED__ = true;
          Object.defineProperty(descriptor.value, "name", {
            value: handlerName,
            configurable: true
          });
          return decoratorResult(target, propertyKey, descriptor);
        };
      };
    };
  }
}

exports.SentryNestEventInstrumentation = SentryNestEventInstrumentation;
