// @sentry/nestjs/build/cjs/integrations/vendored/instrumentation.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const instrumentation = require('@opentelemetry/instrumentation');
const attributes = require('@sentry/conventions/attributes');
const core = require('@sentry/core');
const AttributeNames = require('./enums/AttributeNames.js');
const NestType = require('./enums/NestType.js');

const PACKAGE_NAME = "@sentry/instrumentation-nestjs-core";
const supportedVersions = [">=4.0.0 <12"];
const _NestInstrumentation = class _NestInstrumentation extends instrumentation.InstrumentationBase {
  constructor(config = {}) {
    super(PACKAGE_NAME, core.SDK_VERSION, config);
  }
  init() {
    const module = new instrumentation.InstrumentationNodeModuleDefinition(_NestInstrumentation.COMPONENT, supportedVersions);
    module.files.push(
      this.getNestFactoryFileInstrumentation(supportedVersions),
      this.getRouterExecutionContextFileInstrumentation(supportedVersions)
    );
    return module;
  }
  getNestFactoryFileInstrumentation(versions) {
    return new instrumentation.InstrumentationNodeModuleFile(
      "@nestjs/core/nest-factory.js",
      versions,
      (moduleExports, moduleVersion) => {
        this.ensureWrapped(
          moduleExports.NestFactoryStatic.prototype,
          "create",
          createWrapNestFactoryCreate(moduleVersion)
        );
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports.NestFactoryStatic.prototype, "create");
      }
    );
  }
  getRouterExecutionContextFileInstrumentation(versions) {
    return new instrumentation.InstrumentationNodeModuleFile(
      "@nestjs/core/router/router-execution-context.js",
      versions,
      (moduleExports, moduleVersion) => {
        this.ensureWrapped(
          moduleExports.RouterExecutionContext.prototype,
          "create",
          createWrapCreateHandler(moduleVersion)
        );
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports.RouterExecutionContext.prototype, "create");
      }
    );
  }
  ensureWrapped(obj, methodName, wrapper) {
    if (instrumentation.isWrapped(obj[methodName])) {
      this._unwrap(obj, methodName);
    }
    this._wrap(obj, methodName, wrapper);
  }
};
_NestInstrumentation.COMPONENT = "@nestjs/core";
_NestInstrumentation.COMMON_ATTRIBUTES = {
  component: _NestInstrumentation.COMPONENT
};
let NestInstrumentation = _NestInstrumentation;
function createWrapNestFactoryCreate(moduleVersion) {
  return function wrapCreate(original) {
    return function createWithTrace(...args) {
      const nestModule = args[0];
      return core.startSpan(
        {
          name: "Create Nest App",
          op: `${NestType.NestType.APP_CREATION}.nestjs`,
          attributes: {
            ...NestInstrumentation.COMMON_ATTRIBUTES,
            [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
            [AttributeNames.AttributeNames.TYPE]: NestType.NestType.APP_CREATION,
            [AttributeNames.AttributeNames.VERSION]: moduleVersion,
            [AttributeNames.AttributeNames.MODULE]: nestModule.name
          }
        },
        () => original.apply(this, args)
      );
    };
  };
}
function createWrapCreateHandler(moduleVersion) {
  return function wrapCreateHandler(original) {
    return function createHandlerWithTrace(...args) {
      const instance = args[0];
      const callback = args[1];
      args[1] = createWrapHandler(moduleVersion, callback);
      const handler = original.apply(this, args);
      const callbackName = callback.name;
      const instanceName = instance.constructor?.name || "UnnamedInstance";
      const spanName = callbackName ? `${instanceName}.${callbackName}` : instanceName;
      return function(...handlerArgs) {
        const req = handlerArgs[0];
        const attributes$1 = {
          ...NestInstrumentation.COMMON_ATTRIBUTES,
          [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
          [AttributeNames.AttributeNames.VERSION]: moduleVersion,
          [AttributeNames.AttributeNames.TYPE]: NestType.NestType.REQUEST_CONTEXT,
          [attributes.HTTP_ROUTE]: req.route?.path || req.routeOptions?.url || req.routerPath,
          [AttributeNames.AttributeNames.CONTROLLER]: instanceName,
          [AttributeNames.AttributeNames.CALLBACK]: callbackName
        };
        attributes$1["http.method"] = req.method;
        attributes$1["http.url"] = req.originalUrl || req.url;
        return core.startSpan(
          { name: spanName, op: `${NestType.NestType.REQUEST_CONTEXT}.nestjs`, attributes: attributes$1 },
          () => handler.apply(this, handlerArgs)
        );
      };
    };
  };
}
function createWrapHandler(moduleVersion, handler) {
  const spanName = handler.name || "anonymous nest handler";
  const attributes = {
    ...NestInstrumentation.COMMON_ATTRIBUTES,
    [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
    [AttributeNames.AttributeNames.VERSION]: moduleVersion,
    [AttributeNames.AttributeNames.TYPE]: NestType.NestType.REQUEST_HANDLER,
    [AttributeNames.AttributeNames.CALLBACK]: handler.name
  };
  const wrappedHandler = function(...args) {
    return core.startSpan(
      { name: spanName, op: `${NestType.NestType.REQUEST_HANDLER}.nestjs`, attributes },
      () => handler.apply(this, args)
    );
  };
  if (handler.name) {
    Object.defineProperty(wrappedHandler, "name", { value: handler.name });
  }
  Reflect.getMetadataKeys(handler).forEach((metadataKey) => {
    Reflect.defineMetadata(metadataKey, Reflect.getMetadata(metadataKey, handler), wrappedHandler);
  });
  return wrappedHandler;
}

exports.NestInstrumentation = NestInstrumentation;
