// @sentry/nestjs/build/esm/integrations/vendored/instrumentation.js
import { InstrumentationBase, InstrumentationNodeModuleDefinition, InstrumentationNodeModuleFile, isWrapped } from '@opentelemetry/instrumentation';
import { HTTP_ROUTE } from '@sentry/conventions/attributes';
import { SDK_VERSION, startSpan, SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN } from '@sentry/core';
import { AttributeNames } from './enums/AttributeNames.js';
import { NestType } from './enums/NestType.js';

const PACKAGE_NAME = "@sentry/instrumentation-nestjs-core";
const supportedVersions = [">=4.0.0 <12"];
const _NestInstrumentation = class _NestInstrumentation extends InstrumentationBase {
  constructor(config = {}) {
    super(PACKAGE_NAME, SDK_VERSION, config);
  }
  init() {
    const module = new InstrumentationNodeModuleDefinition(_NestInstrumentation.COMPONENT, supportedVersions);
    module.files.push(
      this.getNestFactoryFileInstrumentation(supportedVersions),
      this.getRouterExecutionContextFileInstrumentation(supportedVersions)
    );
    return module;
  }
  getNestFactoryFileInstrumentation(versions) {
    return new InstrumentationNodeModuleFile(
      "@nestjs/core/nest-factory.js",
      versions,
      (moduleExports, moduleVersion) => {
        this.ensureWrapped(
          moduleExports.NestFactoryStatic.prototype,
          "create",
          createWrapNestFactoryCreate(moduleVersion)
        );
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports.NestFactoryStatic.prototype, "create");
      }
    );
  }
  getRouterExecutionContextFileInstrumentation(versions) {
    return new InstrumentationNodeModuleFile(
      "@nestjs/core/router/router-execution-context.js",
      versions,
      (moduleExports, moduleVersion) => {
        this.ensureWrapped(
          moduleExports.RouterExecutionContext.prototype,
          "create",
          createWrapCreateHandler(moduleVersion)
        );
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports.RouterExecutionContext.prototype, "create");
      }
    );
  }
  ensureWrapped(obj, methodName, wrapper) {
    if (isWrapped(obj[methodName])) {
      this._unwrap(obj, methodName);
    }
    this._wrap(obj, methodName, wrapper);
  }
};
_NestInstrumentation.COMPONENT = "@nestjs/core";
_NestInstrumentation.COMMON_ATTRIBUTES = {
  component: _NestInstrumentation.COMPONENT
};
let NestInstrumentation = _NestInstrumentation;
function createWrapNestFactoryCreate(moduleVersion) {
  return function wrapCreate(original) {
    return function createWithTrace(...args) {
      const nestModule = args[0];
      return startSpan(
        {
          name: "Create Nest App",
          op: `${NestType.APP_CREATION}.nestjs`,
          attributes: {
            ...NestInstrumentation.COMMON_ATTRIBUTES,
            [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
            [AttributeNames.TYPE]: NestType.APP_CREATION,
            [AttributeNames.VERSION]: moduleVersion,
            [AttributeNames.MODULE]: nestModule.name
          }
        },
        () => original.apply(this, args)
      );
    };
  };
}
function createWrapCreateHandler(moduleVersion) {
  return function wrapCreateHandler(original) {
    return function createHandlerWithTrace(...args) {
      const instance = args[0];
      const callback = args[1];
      args[1] = createWrapHandler(moduleVersion, callback);
      const handler = original.apply(this, args);
      const callbackName = callback.name;
      const instanceName = instance.constructor?.name || "UnnamedInstance";
      const spanName = callbackName ? `${instanceName}.${callbackName}` : instanceName;
      return function(...handlerArgs) {
        const req = handlerArgs[0];
        const attributes = {
          ...NestInstrumentation.COMMON_ATTRIBUTES,
          [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
          [AttributeNames.VERSION]: moduleVersion,
          [AttributeNames.TYPE]: NestType.REQUEST_CONTEXT,
          [HTTP_ROUTE]: req.route?.path || req.routeOptions?.url || req.routerPath,
          [AttributeNames.CONTROLLER]: instanceName,
          [AttributeNames.CALLBACK]: callbackName
        };
        attributes["http.method"] = req.method;
        attributes["http.url"] = req.originalUrl || req.url;
        return startSpan(
          { name: spanName, op: `${NestType.REQUEST_CONTEXT}.nestjs`, attributes },
          () => handler.apply(this, handlerArgs)
        );
      };
    };
  };
}
function createWrapHandler(moduleVersion, handler) {
  const spanName = handler.name || "anonymous nest handler";
  const attributes = {
    ...NestInstrumentation.COMMON_ATTRIBUTES,
    [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.http.otel.nestjs",
    [AttributeNames.VERSION]: moduleVersion,
    [AttributeNames.TYPE]: NestType.REQUEST_HANDLER,
    [AttributeNames.CALLBACK]: handler.name
  };
  const wrappedHandler = function(...args) {
    return startSpan(
      { name: spanName, op: `${NestType.REQUEST_HANDLER}.nestjs`, attributes },
      () => handler.apply(this, args)
    );
  };
  if (handler.name) {
    Object.defineProperty(wrappedHandler, "name", { value: handler.name });
  }
  Reflect.getMetadataKeys(handler).forEach((metadataKey) => {
    Reflect.defineMetadata(metadataKey, Reflect.getMetadata(metadataKey, handler), wrappedHandler);
  });
  return wrappedHandler;
}

export { NestInstrumentation };
