// @sentry/nestjs/build/esm/integrations/sentry-nest-bullmq-instrumentation.js
import { InstrumentationBase, InstrumentationNodeModuleDefinition, InstrumentationNodeModuleFile, isWrapped } from '@opentelemetry/instrumentation';
import { SDK_VERSION, withIsolationScope, startSpan, captureException } from '@sentry/core';
import { getBullMQProcessSpanOptions } from './helpers.js';

const supportedVersions = [">=10.0.0"];
const COMPONENT = "@nestjs/bullmq";
class SentryNestBullMQInstrumentation extends InstrumentationBase {
  constructor(config = {}) {
    super("sentry-nestjs-bullmq", SDK_VERSION, config);
  }
  /**
   * Initializes the instrumentation by defining the modules to be patched.
   */
  init() {
    const moduleDef = new InstrumentationNodeModuleDefinition(COMPONENT, supportedVersions);
    moduleDef.files.push(this._getProcessorFileInstrumentation(supportedVersions));
    return moduleDef;
  }
  /**
   * Wraps the @Processor decorator.
   */
  _getProcessorFileInstrumentation(versions) {
    return new InstrumentationNodeModuleFile(
      "@nestjs/bullmq/dist/decorators/processor.decorator.js",
      versions,
      (moduleExports) => {
        if (isWrapped(moduleExports.Processor)) {
          this._unwrap(moduleExports, "Processor");
        }
        this._wrap(moduleExports, "Processor", this._createWrapProcessor());
        return moduleExports;
      },
      (moduleExports) => {
        this._unwrap(moduleExports, "Processor");
      }
    );
  }
  /**
   * Creates a wrapper function for the @Processor class decorator.
   */
  _createWrapProcessor() {
    return function wrapProcessor(original) {
      return function wrappedProcessor(...decoratorArgs) {
        const queueName = (
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          typeof decoratorArgs[0] === "string" ? decoratorArgs[0] : decoratorArgs[0]?.name || "unknown"
        );
        const classDecorator = original(...decoratorArgs);
        return function(target) {
          const originalProcess = target.prototype.process;
          if (originalProcess && typeof originalProcess === "function" && !target.__SENTRY_INTERNAL__ && !originalProcess.__SENTRY_INSTRUMENTED__) {
            target.prototype.process = new Proxy(originalProcess, {
              apply: (originalProcessFn, thisArg, args) => {
                return withIsolationScope(() => {
                  return startSpan(getBullMQProcessSpanOptions(queueName), async () => {
                    try {
                      return await originalProcessFn.apply(thisArg, args);
                    } catch (error) {
                      captureException(error, {
                        mechanism: {
                          handled: false,
                          type: "auto.queue.nestjs.bullmq"
                        }
                      });
                      throw error;
                    }
                  });
                });
              }
            });
            target.prototype.process.__SENTRY_INSTRUMENTED__ = true;
          }
          return classDecorator(target);
        };
      };
    };
  }
}

export { SentryNestBullMQInstrumentation };
