// @sentry/nestjs/build/cjs/decorators.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@sentry/core');
const Sentry = require('@sentry/node');
const helpers = require('./helpers.js');

const SentryCron = (monitorSlug, monitorConfig) => {
  return (target, propertyKey, descriptor) => {
    const originalMethod = descriptor.value;
    descriptor.value = function(...args) {
      return Sentry.withMonitor(
        monitorSlug,
        () => {
          return originalMethod.apply(this, args);
        },
        monitorConfig
      );
    };
    copyFunctionNameAndMetadata({ originalMethod, descriptor });
    return descriptor;
  };
};
function SentryTraced(op = "function") {
  return function(_target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function(...args) {
      return Sentry.startSpan(
        {
          op,
          name: propertyKey,
          attributes: {
            [core.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: "auto.function.nestjs.sentry_traced",
            [core.SEMANTIC_ATTRIBUTE_SENTRY_OP]: op
          }
        },
        () => {
          return originalMethod.apply(this, args);
        }
      );
    };
    copyFunctionNameAndMetadata({ originalMethod, descriptor });
    return descriptor;
  };
}
function SentryExceptionCaptured() {
  return function(target, propertyKey, descriptor) {
    const originalCatch = descriptor.value;
    descriptor.value = function(exception, host, ...args) {
      if (helpers.isExpectedError(exception)) {
        return originalCatch.apply(this, [exception, host, ...args]);
      }
      core.captureException(exception, { mechanism: { handled: false, type: "auto.function.nestjs.exception_captured" } });
      return originalCatch.apply(this, [exception, host, ...args]);
    };
    copyFunctionNameAndMetadata({ originalMethod: originalCatch, descriptor });
    return descriptor;
  };
}
function copyFunctionNameAndMetadata({
  originalMethod,
  descriptor
}) {
  Object.defineProperty(descriptor.value, "name", {
    value: originalMethod.name,
    configurable: true,
    enumerable: true,
    writable: true
  });
  if (typeof Reflect !== "undefined" && typeof Reflect.getMetadataKeys === "function") {
    const originalMetaData = Reflect.getMetadataKeys(originalMethod);
    for (const key of originalMetaData) {
      const value = Reflect.getMetadata(key, originalMethod);
      Reflect.defineMetadata(key, value, descriptor.value);
    }
  }
}

exports.SentryCron = SentryCron;
exports.SentryExceptionCaptured = SentryExceptionCaptured;
exports.SentryTraced = SentryTraced;
