// @sentry/nestjs/build/cjs/setup.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const common = require('@nestjs/common');
const core = require('@nestjs/core');
const core$1 = require('@sentry/core');
const helpers = require('./helpers.js');

class SentryTracingInterceptor {
  constructor() {
    this.__SENTRY_INTERNAL__ = true;
  }
  /**
   * Intercepts HTTP requests to set the transaction name for Sentry tracing.
   */
  intercept(context, next) {
    if (core$1.getIsolationScope() === core$1.getDefaultIsolationScope()) {
      core$1.debug.warn("Isolation scope is still the default isolation scope, skipping setting transactionName.");
      return next.handle();
    }
    if (context.getType() === "http") {
      const req = context.switchToHttp().getRequest();
      if ("routeOptions" in req && req.routeOptions?.url) {
        core$1.getIsolationScope().setTransactionName(`${(req.method || "GET").toUpperCase()} ${req.routeOptions.url}`);
      } else if ("route" in req && req.route?.path) {
        core$1.getIsolationScope().setTransactionName(`${(req.method || "GET").toUpperCase()} ${req.route.path}`);
      }
    }
    return next.handle();
  }
}
common.Injectable()(SentryTracingInterceptor);
class SentryGlobalFilter extends core.BaseExceptionFilter {
  constructor(applicationRef) {
    super(applicationRef);
    this.__SENTRY_INTERNAL__ = true;
    this._logger = new common.Logger("ExceptionsHandler");
  }
  /**
   * Catches exceptions and reports them to Sentry unless they are expected errors.
   */
  catch(exception, host) {
    const contextType = host.getType();
    if (contextType === "graphql") {
      if (exception instanceof common.HttpException) {
        throw exception;
      }
      if (exception instanceof Error) {
        this._logger.error(exception.message, exception.stack);
      }
      core$1.captureException(exception, {
        mechanism: {
          handled: false,
          type: "auto.graphql.nestjs.global_filter"
        }
      });
      throw exception;
    }
    if (contextType === "rpc") {
      if (exception instanceof common.HttpException) {
        throw exception;
      }
      if (!(exception instanceof Error)) {
        if (!helpers.isExpectedError(exception)) {
          core$1.captureException(exception, {
            mechanism: {
              handled: false,
              type: "auto.rpc.nestjs.global_filter"
            }
          });
        }
        throw exception;
      }
      if (!helpers.isExpectedError(exception)) {
        core$1.captureException(exception, {
          mechanism: {
            handled: false,
            type: "auto.rpc.nestjs.global_filter"
          }
        });
      }
      this._logger.warn(
        "IMPORTANT: RpcException should be handled with a dedicated Rpc exception filter, not the generic SentryGlobalFilter"
      );
      this._logger.error(exception.message, exception.stack);
      return;
    }
    if (!helpers.isExpectedError(exception)) {
      core$1.captureException(exception, {
        mechanism: {
          handled: false,
          type: "auto.http.nestjs.global_filter"
        }
      });
    }
    return super.catch(exception, host);
  }
}
common.Catch()(SentryGlobalFilter);
class SentryModule {
  /**
   * Configures the module as the root module in a Nest.js application.
   */
  static forRoot() {
    return {
      module: SentryModule,
      providers: [
        {
          provide: core.APP_INTERCEPTOR,
          useClass: SentryTracingInterceptor
        }
      ]
    };
  }
}
common.Global()(SentryModule);
common.Module({
  providers: [
    {
      provide: core.APP_INTERCEPTOR,
      useClass: SentryTracingInterceptor
    }
  ]
})(SentryModule);

exports.SentryGlobalFilter = SentryGlobalFilter;
exports.SentryModule = SentryModule;
