// @sentry/server-utils/build/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const redisDcSubscriber = require('./redis/redis-dc-subscriber.js');
const tracingChannel = require('./tracing-channel.js');
const index$1 = require('./vercel-ai/index.js');
const index = require('./integrations/tracing-channel/fastify/index.js');



exports.IOREDIS_DC_CHANNEL_COMMAND = redisDcSubscriber.IOREDIS_DC_CHANNEL_COMMAND;
exports.IOREDIS_DC_CHANNEL_CONNECT = redisDcSubscriber.IOREDIS_DC_CHANNEL_CONNECT;
exports.REDIS_DC_CHANNEL_BATCH = redisDcSubscriber.REDIS_DC_CHANNEL_BATCH;
exports.REDIS_DC_CHANNEL_COMMAND = redisDcSubscriber.REDIS_DC_CHANNEL_COMMAND;
exports.REDIS_DC_CHANNEL_CONNECT = redisDcSubscriber.REDIS_DC_CHANNEL_CONNECT;
exports.subscribeRedisDiagnosticChannels = redisDcSubscriber.subscribeRedisDiagnosticChannels;
exports.bindTracingChannelToSpan = tracingChannel.bindTracingChannelToSpan;
exports.vercelAiIntegration = index$1.vercelAiIntegration;
exports.fastifyIntegration = index.fastifyIntegration;
exports.handleFastifyError = index.handleFastifyError;
exports.instrumentFastify = index.instrumentFastify;
